ospf: Add RFC 9339 support

nrybowski opened the pull request:
This PR adds the support for Reverse Metric and Reverse TE Metric advertisement through the LLS data block of Hello packets, as defined by RFC 9339.

It depends on RFC 5613 whose support has been added with https://github.com/holo-routing/holo/pull/77.

### TODO
- [x] Add Reverse (TE) Metric LLS TLVs encoding / decoding
    - [x] Add encoding / decoding tests for OSPFv2
    - [x] Add encoding / decoding tests for OSPFv3
- [x] Enable Reverse Metric configuration
    - [x] Add YANG model. (There is no YANG model specific to OSPF for now, I will adapt the one for IS-IS.)
    - [x] Extend northbound interface
         - [x] Directly send Hello packet upon Reverse Metric enabling / disabling
    - [x] Extend OSPF configuration to store Reverse Metric configuration
- [x] Implement Reverse Metric logic
    - [x] Update metric if needed upon Reverse Metric reception
    - [x] Trigger Router LSA origination upon Reverse Metric reception
    - [ ] Fallback to provisioned metric once Reverse Metric is not advertised anymore by the neighbor.
- [ ] Enable Reverse TE Metric configuration
    - [ ] Add YANG model.
    - [ ] Extend northbound interface
         - [ ] Directly send Hello packet upon Reverse TE Metric enabling / disabling
    - [ ] Extend OSPF configuration to store Reverse TE Metric configuration
- [ ] Implement Reverse TE Metric logic
    - [ ] Update metric if needed upon Reverse TE Metric reception
    - [ ] Trigger Router LSA origination upon Reverse TE Metric reception
    - [ ] Fallback to provisioned metric once Reverse TE Metric is not advertised anymore by the neighbor.

nrybowski committed:
ospf: Add support for RFC 9339

This patch adds Reverse Metric and Reverse TE Metric TLVs to LLS data
blocks.

This patch hence depends on RFC 5613, added with
https://github.com/holo-routing/holo/pull/77.

Signed-off-by: Nicolas Rybowski <nicolas.rybowski@uclouvain.be>

nrybowski committed:
ospfv2: Encode/decode tests for Reverse(TE) Metric

Signed-off-by: Nicolas Rybowski <nicolas.rybowski@uclouvain.be>

nrybowski committed:
ospfv3: Encode/decode tests for Reverse(TE) Metric

Signed-off-by: Nicolas Rybowski <nicolas.rybowski@uclouvain.be>

nrybowski committed:
ospf: Handle Reverse Metric.

This commit adds a custom YANG module to configure
Reverse Metrics per interface and the related callbacks.
It also implements the handling of such Reverse Metric
both at the sender and receiver sides.

Signed-off-by: Nicolas Rybowski <nicolas.rybowski@uclouvain.be>

nrybowski commented on holo-ospf/src/events.rs:
```diff
@@ -381,9 +383,63 @@ where
+                    if cfg.higher && iface.config.cost < cfg.metric
+                        || !cfg.higher
+                    {
+                        if cfg.offset {
+                            iface.config.cost =
+                                iface.config.cost.saturating_add(cfg.metric);
+                        } else {
+                            iface.config.cost = cfg.metric;
+                        }
```
This is not correct. We can not change the cost loaded from the configuration. We should rather extend the `InterfaceState` with the current cost value so that we can fallback later on the provisioned cost.

nrybowski replied:
Fixed in https://github.com/holo-routing/holo/commit/46ab06ea67338b611f1cd149eb64b924931f16ce.

rwestphal reviewed (commented):
Awesome work. I really like how this is shaping up!

As a preliminary review comment, I think the YANG module could be simplified.

Currently, the interface augmentation looks like this:
```
     +--rw holo-ospf-reverse-metric:reverse-metric
        +--rw holo-ospf-reverse-metric:enable-receive?     boolean
        +--rw holo-ospf-reverse-metric:enable-advertise?   boolean
        +--rw holo-ospf-reverse-metric:config* [mtid]
           +--rw holo-ospf-reverse-metric:mtid      uint8
           +--rw holo-ospf-reverse-metric:metric?   ospf:ospf-link-metric
           +--rw holo-ospf-reverse-metric:flags
              +--rw holo-ospf-reverse-metric:higher?   boolean
              +--rw holo-ospf-reverse-metric:offset?   boolean
```

It could be simplified to something like this, more in line with the IS-IS RM module:
```
     +--rw holo-ospf-reverse-metric:reverse-metric
        +--rw holo-ospf-reverse-metric:enable-receive?     boolean
        +--rw holo-ospf-reverse-metric:metric?   ospf:ospf-link-metric
        +--rw holo-ospf-reverse-metric:flags
          +--rw holo-ospf-reverse-metric:higher?   boolean
          +--rw holo-ospf-reverse-metric:offset?   boolean
```

Later, once support for RFC 4915 is implemented, a separate augmentation could be added for "interface/topologies/topology". I think that would be cleaner, especially since "mtid" only makes sense for OSPFv2.

Also, the "enable-advertise" leaf can be removed, as the presence of "metric" already implies we want to advertise the reverse metric.

What do you think?

rwestphal commented on holo-ospf/src/events.rs:
```diff
@@ -381,9 +383,63 @@ where
+            // If the Reverse Metric from the neighbor has been updated since
+            // the last one we received, then we propagate the information.
+            for (mtid, (flags, metric)) in lls.reverse_metric.iter() {
+                let entry = iface.config.reverse_metric.rm_rx.entry(*mtid);
```
Indeed. We can probably keep a copy of the last received RM TLV in `InstanceState` to facilitate checking for changes.

nrybowski replied:
Fixed in https://github.com/holo-routing/holo/pull/78/commits/d0236c0bbd4670b35ea070205a8c2fa2db72107f.

nrybowski committed:
ospf: Simplify Reverse Metric YANG model.

This commit addresses the comment from https://github.com/holo-routing/holo/pull/78#pullrequestreview-3295787903
with the suggestion proposed in https://github.com/holo-routing/holo/pull/78#issuecomment-3363641258.

Signed-off-by: Nicolas Rybowski <nicolas.rybowski@uclouvain.be>

nrybowski commented:
Just rebased on [master](49a2d4ebde079164a872784c212071101a993797) to clear the conflicts. https://github.com/holo-routing/holo/pull/78/commits/81df85f61627aa0f60524e0323a88ab17984b1ec contains the changes we just discussed.

> Just keep in mind that in "interface/topologies/topology" you can configure a topology name, but not an "mtid", which can be a bit confusing. This is probably because the IETF folks couldn't agree on whether the MT-ID should be explicitly configured or inferred from the selected RIB. In any case, this shouldn't affect your augmentation. I just wanted to point it out in case it causes any confusion :)

The good news is that the "mtid" augmentation for OSPFv2 is gated with the feature "multi-topology" so we don't even need a deviation to "unsupport" the augmentation and we can keep the headache of handling that when RFC 4915 is implemented :)

rwestphal commented:
> The good news is that the "mtid" augmentation for OSPFv2 is gated with the feature "multi-topology" so we don't even need a deviation to "unsupport" the augmentation and we can keep the headache of handling that when RFC 4915 is implemented :)

Yeah that makes things easier for us :)

From a quick look, it seems most of the RFC is already implemented. Tomorrow I'll run some tests and do a full review. Looking forward to it :)

rwestphal reviewed (commented):
I just ran some tests and it's very interesting to see the reverse metric in action! It's quite a useful feature for operators to have in their toolbox.

Please find my review comments below.
